Fix submission type #171270123

software-project opened the pull request:
Fix for [#171270123](https://www.pivotaltracker.com/n/projects/1405352/stories/171270123) and https://www.pivotaltracker.com/story/show/172804251

software-project committed:
Fix submission type #171270123

software-project committed:
Merge branch 'master' into bugs/fix-submission-type-#171270123

akegan reviewed (commented):
In other cases I thought we were able to distinguish between the value we display and the actual value of the filter. Would that be possible here?

akegan committed:
Merge branch 'master' into bugs/fix-submission-type-#171270123

software-project commented:
Well, I just found another bug. Select a listing, click filter, change to any listing and then filter again. We have two ways of fixing it I suppose. One in the controller or in React. Which would you prefer?

james-wills-sf commented:
I'm going to take over this PR since Martin is out of hours for the day.

james-wills-sf committed:
Fix "Any Listing" + "Any Type" options #172804251

We were sending "Any Listing" and "Any Type" as selection values to the
backend. This is not the expected behavior, as those are labels, not
values, and their values should be null. The issue was the way the form
fields render with null values, they pass the label as the value on
submit. This change makes it so that we always pass null when the value
is null.

Additionally, we add filtering to scrub requests with empty strings as
values, because on this page all empty strings mean "don't filter".

james-wills-sf commented:
@akegan @software-project updated this to fix the issue on the frontend.

It turns out, when we used "null" as a value for one option, the actual input created looks like this:
```
<select ...>
  <option>Any Listing</option>
  <option value="<listing-id>">Some listing name</option>
  ...
</select>
```

And when the first option is selected on submit it uses the label, "Any Listing", since there is no value attr. I fixed it by passing in an empty string as value instead of null, so the html looks like this instead:
```
<select ...>
  <option value="">Any Listing</option>
  <option value="<listing-id>">Some listing name</option>
  ...
</select>
```

And then I scrub the filters for empty strings before passing them to the api.

One other nice thing this fixes is before we would pass filters like `application-name: ""` if you typed something in that text field then deleted it. The scrubber works on all fields so that won't happen anymore.

![applicationfilter](https://user-images.githubusercontent.com/64036574/81746708-cc600100-945b-11ea-9d11-c85d68358232.gif)

james-wills-sf commented on spec/javascript/components/applications/__snapshots__/ListingApplicationsPage.test.js.snap:
```diff
@@ -201,7 +201,7 @@ Array [
                   <option
                     value=""
                   >
-                    Select One...
```
The type field used to default to "Select One..." with "Any Type" at the bottom of the list. I considered this a bug since select one and any type have the same functionality, so now we just default to "Any Type"

## Screenshot (before)
<img width="223" alt="- 2020-05-12 at 2 24 29 PM" src="https://user-images.githubusercontent.com/64036574/81747174-78a1e780-945c-11ea-8c39-074eec665bc5.png">

## Screenshot (after)
<img width="224" alt="- 2020-05-12 at 2 24 10 PM" src="https://user-images.githubusercontent.com/64036574/81747217-8ce5e480-945c-11ea-9465-6658ea3b8cfe.png">

akegan reviewed (approved):
These changes look good. Just noticed that we have the same bug in our lease up applications table too. Not sure if it's worth fixing in this PR too or we could make a separate ticket, up to you

james-wills-sf committed:
Fix select fields for lease up application filter

james-wills-sf committed:
update snapshot

james-wills-sf commented:
@akegan I went ahead and fixed lease up too.

I also did a code search to see if this is happening anywhere else and I only found one place in [ConfirmedHouseholdIncome.js](https://github.com/SFDigitalServices/sf-dahlia-lap/blob/master/app/javascript/components/supplemental_application/sections/ConfirmedHouseholdIncome.js#L47), but it appears that they're manually handling select changes so I am not going to mess with it.

![leaseupapplicationfilter](https://user-images.githubusercontent.com/64036574/81751869-2addad00-9465-11ea-8d7d-d461d9442783.gif)

akegan reviewed (approved):
looks good! thanks for making the update
